pkg/codegen/hcl2/model/type_object.go

```diff
@@ -110,6 +111,19 @@ func (t *ObjectType) equals(other Type, seen map[Type]struct{}) bool {
 			return false
 		}
 	}
+	if len(t.Annotations) != len(otherObject.Annotations) {
```

Comment:
Some discriminated union types have a shape `A | B` where both `A` and `B` have exactly the same property names. Currently, these types are smashed into one when a union is built, which is wrong.

If we don't want to account for annotations in the equality check, we'd probably need to re-introduce the `Const` type from my other PR, because that's the only difference between two types.